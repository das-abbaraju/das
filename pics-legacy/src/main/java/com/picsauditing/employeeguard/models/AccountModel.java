package com.picsauditing.employeeguard.models;

public final class AccountModel implements Comparable<AccountModel> {

    private final int id;
    private final String name;
    private final AccountType accountType;

    public AccountModel(Builder builder) {
        this.id = builder.id;
        this.name = builder.name;
        this.accountType = builder.accountType;
    }

    public int getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public AccountType getAccountType() {
        return accountType;
    }

    public boolean isCorporate() {
        return accountType == AccountType.CORPORATE;
    }

    @Override
    public int compareTo(AccountModel that) {
        return this.name.compareToIgnoreCase(that.name);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        AccountModel that = (AccountModel) o;

        if (id != that.id) return false;
        if (accountType != that.accountType) return false;
        if (name != null ? !name.equals(that.name) : that.name != null) return false;

        return true;
    }

    @Override
    public int hashCode() {
        int result = id;
        result = 31 * result + (name != null ? name.hashCode() : 0);
        result = 31 * result + (accountType != null ? accountType.hashCode() : 0);
        return result;
    }

    public static class Builder {

        private int id;
        private String name;
        private AccountType accountType;

        public Builder id(int id) {
            this.id = id;
            return this;
        }

        public Builder name(String name) {
            this.name = name;
            return this;
        }

        public Builder accountType(AccountType accountType) {
            this.accountType = accountType;
            return this;
        }

        public AccountModel build() {
            return new AccountModel(this);
        }
    }
}
