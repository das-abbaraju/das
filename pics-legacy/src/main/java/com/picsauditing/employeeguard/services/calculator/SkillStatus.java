package com.picsauditing.employeeguard.services.calculator;

public enum SkillStatus {

    // TODO: Rename expiring to something more meaningful that the Skill is about to expire

    // Keep in order of serverity, from highest severity to lowest
    Expired("expired"),
    Expiring("expiring"),
    Pending("pending"),
    Complete("complete");

    private String displayValue;

    private SkillStatus(String displayValue) {
        this.displayValue = displayValue;
    }

    public boolean isExpired() {
        return this == Expired;
    }

    public boolean isExpiring() {
        return this == Expiring;
    }

    public boolean isPending() {
        return this == Pending;
    }

    public boolean isComplete() {
        return this == Complete;
    }

    public String getDisplayValue() {
        return displayValue;
    }
}
