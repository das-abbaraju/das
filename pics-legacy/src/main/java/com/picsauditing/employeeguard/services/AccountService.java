package com.picsauditing.employeeguard.services;

import com.picsauditing.PICS.Utilities;
import com.picsauditing.dao.AccountDAO;
import com.picsauditing.dao.ContractorAccountDAO;
import com.picsauditing.dao.OperatorAccountDAO;
import com.picsauditing.employeeguard.daos.AccountEmployeeGuardDAO;
import com.picsauditing.employeeguard.entities.Employee;
import com.picsauditing.employeeguard.entities.Profile;
import com.picsauditing.employeeguard.services.external.BillingService;
import com.picsauditing.employeeguard.services.models.AccountModel;
import com.picsauditing.employeeguard.services.models.AccountType;
import com.picsauditing.employeeguard.util.Extractor;
import com.picsauditing.employeeguard.util.ExtractorUtil;
import com.picsauditing.jpa.entities.Account;
import com.picsauditing.jpa.entities.ContractorAccount;
import com.picsauditing.jpa.entities.ContractorOperator;
import com.picsauditing.jpa.entities.OperatorAccount;
import com.picsauditing.provisioning.ProductSubscriptionService;
import org.apache.commons.collections.CollectionUtils;
import org.apache.commons.lang3.ArrayUtils;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.*;

/**
 * This is a stand-in for a "remote" service call
 */
public class AccountService {

	@Autowired
	private AccountDAO accountDAO;
	@Autowired
	private AccountEmployeeGuardDAO accountEmployeeGuardDAO;
	@Autowired
	private ContractorAccountDAO contractorDAO;
	@Autowired
	private OperatorAccountDAO operatorDAO;
	@Autowired
	private BillingService billingService;
	@Autowired
	private ProductSubscriptionService productSubscriptionService;

	public AccountModel getAccountById(int accountId) {
		Account account = accountDAO.find(accountId);
		return mapAccountToAccountModel(account);
	}

	public List<AccountModel> getAccountsByIds(Collection<Integer> accountIds) {
		List<Account> accounts = accountDAO.findByIds(accountIds);
		return mapAccountsToAccountModels(accounts);
	}

	public List<AccountModel> getTopmostCorporateAccounts(final int accountId) {
		if (accountId <= 0) {
			throw new IllegalArgumentException("Invalid account ID: " + accountId);
		}

		List<OperatorAccount> employeeGUARDCorporates = getEmployeeGUARDCorporates(Arrays.asList(accountId));
		// We don't have a need to modify accounts, so we'll map these corporate accounts to AccountModels
		return mapAccountsToAccountModels(employeeGUARDCorporates);
	}

	public Map<Integer, AccountModel> getIdToAccountModelMap(final Collection<Integer> accountIds) {
		return Utilities.convertToMap(getAccountsByIds(accountIds), new Utilities.MapConvertable<Integer, AccountModel>() {

			@Override
			public Integer getKey(AccountModel accountModel) {
				return accountModel.getId();
			}
		});
	}

	public List<Integer> getTopmostCorporateAccountIds(final Collection<Integer> accountIds) {
		if (CollectionUtils.isEmpty(accountIds)) {
			return Collections.emptyList();
		}

		return extractIdFromAccountModel(getEmployeeGUARDCorporates(accountIds));
	}

	public List<Integer> getTopmostCorporateAccountIds(final int accountId) {
		if (accountId <= 0) {
			throw new IllegalArgumentException("Invalid account ID: " + accountId);
		}

		return extractIdFromAccountModel(getEmployeeGUARDCorporates(Arrays.asList(accountId)));
	}

	private List<OperatorAccount> getEmployeeGUARDCorporates(Collection<Integer> accountIds) {
		List<OperatorAccount> operators = operatorDAO.findOperators(new ArrayList<>(accountIds));

		ArrayList<Integer> visited = new ArrayList<>();

		Set<OperatorAccount> corporates = new HashSet<>();
		for (OperatorAccount operator : operators) {
			List<OperatorAccount> topmostCorporates = getTopmostCorporates(operator, visited);
			corporates.addAll(billingService.filterEmployeeGUARDAccounts(topmostCorporates));
		}

		return new ArrayList<>(corporates);
	}

	private List<OperatorAccount> getTopmostCorporates(OperatorAccount operator, List<Integer> visited) {
		List<OperatorAccount> corporates = new ArrayList<>();
		if (onlyHasPicsConsortiumOrNoParents(operator)) {
			visited.add(operator.getId());
			corporates.add(operator);
		} else {
			for (OperatorAccount parent : operator.getParentOperators()) {
				if (!parent.isInPicsConsortium() && !visited.contains(parent.getId())) {
					visited.add(parent.getId());
					corporates.addAll(getTopmostCorporates(parent, visited));
				}
			}
		}

		return corporates;
	}

	private boolean onlyHasPicsConsortiumOrNoParents(OperatorAccount operator) {
		if (operator.getParentOperators().size() == 0) {
			return true;
		}

		for (OperatorAccount parent : operator.getParentOperators()) {
			if (!parent.isInPicsConsortium()) {
				return false;
			}
		}

		return true;
	}

	public List<AccountModel> getChildOperators(final int accountId) {
		return getChildOperators(Arrays.asList(accountId));
	}

	public List<AccountModel> getChildOperators(final List<Integer> accountIds) {
		List<OperatorAccount> corporates = operatorDAO.findOperators(accountIds);
		if (CollectionUtils.isEmpty(corporates)) {
			return Collections.emptyList();
		}

		Set<OperatorAccount> childAccounts = new HashSet<>();
		for (OperatorAccount corporate : corporates) {
			List<OperatorAccount> childOperators = new ArrayList<>(corporate.getChildOperators());
			childAccounts.addAll(billingService.filterEmployeeGUARDAccounts(childOperators));
		}

		return mapAccountsToAccountModels(new ArrayList<>(childAccounts));
	}

	public List<Integer> getChildOperatorIds(final int accountId) {
		return extractIdFromAccountModel(getChildOperators(accountId).toArray(new AccountModel[0]));
	}

	public List<Integer> getChildOperatorIds(final List<Integer> accountIds) {
		return extractIdFromAccountModel(getChildOperators(accountIds).toArray(new AccountModel[0]));
	}

	public AccountType getAccountTypeByUserID(int userID) {
		return getAccountByUserID(userID).getAccountType();
	}

	public AccountModel getAccountByUserID(int userID) {
		return getAccountById(accountDAO.findByUserID(userID));
	}

	public Map<Integer, AccountModel> getContractorMapForSite(final int siteId) {
		return Utilities.convertToMap(getContractors(siteId), new Utilities.MapConvertable<Integer, AccountModel>() {
			@Override
			public Integer getKey(AccountModel accountModel) {
				return accountModel.getId();
			}
		});
	}

	/**
	 * Retrieve the list of Contractor IDs that work under this Operator/Corporate hierarchy and have EmployeeGUARD.
	 *
	 * @param siteId The ID of the Operator Facility/Site or the ID of the Corporation
	 * @return
	 */
	public List<Integer> getContractorIds(final int siteId) {
		return extractIdFromAccountModel(getContractors(siteId).toArray(new AccountModel[0]));
	}

	public List<Integer> getContractorIds(final Collection<Integer> siteIds) {
		if (CollectionUtils.isEmpty(siteIds)) {
			return Collections.emptyList();
		}

		return contractorDAO.findAllContractorIdsForOperatorIds(siteIds);
	}

	/**
	 * Retrieve the list of Contractors that work under this Operator/Corporate hierarchy and have EmployeeGUARD.
	 *
	 * @param siteId The ID of the Operator Facility/Site or the ID of the Corporation
	 * @return
	 */
	public List<AccountModel> getContractors(final int siteId) {
		OperatorAccount operator = operatorDAO.find(siteId);
		if (operator == null) {
			return Collections.emptyList();
		}

		List<ContractorAccount> contractors = new ArrayList<>();
		if (!operator.isCorporate()) {
			contractors = contractorDAO.findAllContractorsForOperator(operator);
		} else {
			for (OperatorAccount site : operator.getChildOperators()) {
				addContractorsFromOperator(site, contractors);
			}
		}

		List<ContractorAccount> accounts = billingService.filterEmployeeGUARDAccounts(contractors);

		return mapAccountsToAccountModels(accounts);
	}

	private void addContractorsFromOperator(OperatorAccount operator, List<ContractorAccount> contractors) {
		for (ContractorOperator contractorOperator : operator.getContractorOperators()) {
			contractors.add(contractorOperator.getContractorAccount());
		}
	}

	private <E extends Account> List<AccountModel> mapAccountsToAccountModels(List<E> accounts) {
		if (CollectionUtils.isEmpty(accounts)) {
			return Collections.emptyList();
		}

		List<AccountModel> accountModels = new ArrayList<>(accounts.size());
		for (Account account : accounts) {
			accountModels.add(mapAccountToAccountModel(account));
		}

		return accountModels;
	}

	private AccountModel mapAccountToAccountModel(Account account) {
		return new AccountModel.Builder().accountType(getAccountTypeForAccount(account)).id(account.getId())
				.name(account.getName()).build();
	}

	private AccountType getAccountTypeForAccount(Account account) {
		switch (account.getType()) {
			case "Admin":
				return AccountType.ADMIN_ACCOUNT;

			case "Assessment":
				return AccountType.ASSESSMENT;

			case "Contractor":
				return AccountType.CONTRACTOR;

			case "Corporate":
				return AccountType.CORPORATE;

			case "Operator":
				return AccountType.OPERATOR;

			default:
				throw new IllegalArgumentException("Invalid account type " + account.getType());
		}
	}

	public List<Integer> extractIdFromAccountModel(final AccountModel... accountModels) {
		if (ArrayUtils.isEmpty(accountModels)) {
			return Collections.emptyList();
		}

		List<Integer> ids = new ArrayList<>();
		for (AccountModel accountModel : accountModels) {
			ids.add(accountModel.getId());
		}

		return ids;
	}

	private List<Integer> extractIdFromAccountModel(List<? extends Account> accounts) {
		if (CollectionUtils.isEmpty(accounts)) {
			return Collections.emptyList();
		}

		List<Integer> ids = new ArrayList<>();
		for (Account account : accounts) {
			ids.add(account.getId());
		}

		return ids;
	}

	public boolean doesContractorStillNeedEmployeeGuard(final int contractorID) {
		return doesContractorStillNeedEmployeeGuard(accountDAO.find(contractorID));
	}

	private boolean doesContractorStillNeedEmployeeGuard(final Account account) {
		return productSubscriptionService.hasEmployeeGUARD(account);
	}

	public Map<Integer, AccountModel> getContractorsForEmployee(final Employee employee) {
		if (employee == null) {
			return Collections.emptyMap();
		}

		if (employee.getProfile() != null) {
			return getContractorsForProfile(employee.getProfile());
		}

		return getIdToAccountModelMap(Arrays.asList(employee.getAccountId()));
	}

	public Map<Integer, AccountModel> getContractorsForProfile(final Profile profile) {
		if (profile == null) {
			return Collections.emptyMap();
		}

		List<Integer> contractorIds = ExtractorUtil.extractList(profile.getEmployees(),
				new Extractor<Employee, Integer>() {
					@Override
					public Integer extract(Employee employee) {
						return employee.getAccountId();
					}
				});

		return getIdToAccountModelMap(contractorIds);
	}

	public Map<Integer, AccountModel> getContractorsForEmployeesMap(final List<Employee> employees) {
		return getIdToAccountModelMap(getAccountIds(employees));
	}

	public List<AccountModel> getContractorsForEmployees(final List<Employee> employees) {
		if (CollectionUtils.isEmpty(employees)) {
			return Collections.emptyList();
		}

		return getAccountsByIds(getAccountIds(employees));
	}

	private List<Integer> getAccountIds(final List<Employee> employees) {
		List<Integer> accountIds = new ArrayList<>();
		for (Employee employee : employees) {
			accountIds.add(employee.getAccountId());
		}

		return accountIds;
	}

	public List<Integer> getOperatorIdsForContractor(final int contractorId) {
		ContractorAccount contractor = contractorDAO.find(contractorId);
		if (contractor == null) {
			return Collections.emptyList();
		}

		return operatorDAO.findAllOperatorsForContractor(contractor);
	}
}
