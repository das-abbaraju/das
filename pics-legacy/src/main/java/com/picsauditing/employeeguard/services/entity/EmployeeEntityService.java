package com.picsauditing.employeeguard.services.entity;

import com.picsauditing.PICS.PICSFileType;
import com.picsauditing.PICS.Utilities;
import com.picsauditing.employeeguard.daos.EmployeeDAO;
import com.picsauditing.employeeguard.daos.ProjectRoleEmployeeDAO;
import com.picsauditing.employeeguard.daos.RoleEmployeeDAO;
import com.picsauditing.employeeguard.entities.*;
import com.picsauditing.employeeguard.entities.helper.EntityHelper;
import com.picsauditing.employeeguard.forms.PhotoForm;
import com.picsauditing.employeeguard.models.EntityAuditInfo;
import com.picsauditing.employeeguard.util.PhotoUtil;
import com.picsauditing.util.FileUtils;
import com.picsauditing.util.Strings;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.*;

public class EmployeeEntityService implements EntityService<Employee, Integer>, Searchable<Employee> {

	@Autowired
	private EmployeeDAO employeeDAO;
	@Autowired
	private PhotoUtil photoUtil;
	@Autowired
	private ProjectRoleEmployeeDAO projectRoleEmployeeDAO;
	@Autowired
	private RoleEmployeeDAO roleEmployeeDAO;

	/* All Find Methods */

	@Override
	public Employee find(final Integer id) {
		if (id == null) {
			throw new NullPointerException("id cannot be null");
		}

		return employeeDAO.find(id);
	}

	public Employee find(final int id, final int accountId) {
		return employeeDAO.findEmployeeByAccount(id, accountId);
	}

	public List<Employee> getEmployeesForAccount(final int accountId) {
		return employeeDAO.findByAccount(accountId);
	}

	public long getNumberOfEmployeesForAccount(final int accountId) {
		return employeeDAO.findEmployeeCount(accountId);
	}

	public List<Employee> getEmployeesForAccounts(final Collection<Integer> accountIds) {
		return employeeDAO.findByAccounts(accountIds);
	}

	public List<Employee> getEmployeesForProjects(final List<Project> projects) {
		return employeeDAO.findByProjects(projects);
	}

	public Map<Project, Set<Employee>> getEmployeesByProject(final List<Project> projects) {
		return Utilities.convertToMapOfSets(projectRoleEmployeeDAO.findByProjects(projects),
				new Utilities.EntityKeyValueConvertable<ProjectRoleEmployee, Project, Employee>() {

					@Override
					public Project getKey(ProjectRoleEmployee projectRoleEmployee) {
						return projectRoleEmployee.getProjectRole().getProject();
					}

					@Override
					public Employee getValue(ProjectRoleEmployee projectRoleEmployee) {
						return projectRoleEmployee.getEmployee();
					}
				});
	}

	/**
	 * Retrieves a map of employees to both site and project roles for the specific site
	 *
	 * @param siteId
	 * @param contractorIds
	 * @param projects
	 * @param siteToContractorRoles
	 * @return
	 */
	public Map<Role, Set<Employee>> getEmployeesByRole(final int siteId,
	                                                   final List<Integer> contractorIds,
	                                                   final List<Project> projects,
	                                                   final Map<Role, Role> siteToContractorRoles) {

		Map<Role, Set<Employee>> projectEmployees = getEmployeesByProjectRoles(projects);
		Map<Role, Set<Employee>> siteRoleEmployees = getEmployeesBySiteRoles(siteId, contractorIds, siteToContractorRoles);

		return Utilities.mergeValuesOfMapOfSets(projectEmployees, siteRoleEmployees);
	}

	private Map<Role, Set<Employee>> getEmployeesByProjectRoles(List<Project> projects) {
		return Utilities.convertToMapOfSets(
				projectRoleEmployeeDAO.findByProjects(projects),
				new Utilities.EntityKeyValueConvertable<ProjectRoleEmployee, Role, Employee>() {
					@Override
					public Role getKey(ProjectRoleEmployee entity) {
						return entity.getProjectRole().getRole();
					}

					@Override
					public Employee getValue(ProjectRoleEmployee entity) {
						return entity.getEmployee();
					}
				}
		);
	}

	private Map<Role, Set<Employee>> getEmployeesBySiteRoles(int siteId, List<Integer> contractorIds, final Map<Role, Role> siteToContractorRoles) {
		List<Employee> employeesAssignedToSite = employeeDAO.findEmployeesAssignedToSite(contractorIds, siteId);
		return Utilities.convertToMapOfSets(
				roleEmployeeDAO.findByEmployeesAndSiteId(employeesAssignedToSite, siteId),
				new Utilities.EntityKeyValueConvertable<RoleEmployee, Role, Employee>() {
					@Override
					public Role getKey(RoleEmployee entity) {
						return siteToContractorRoles.get(entity.getRole());
					}

					@Override
					public Employee getValue(RoleEmployee entity) {
						return entity.getEmployee();
					}
				}
		);
	}

	public List<Employee> getEmployeesAssignedToSite(final Collection<Integer> contractorIds, final int siteId) {
		return employeeDAO.findEmployeesAssignedToSite(contractorIds, siteId);
	}

	/* All Search Methods */

	@Override
	public List<Employee> search(final String searchTerm, final int accountId) {
		if (Strings.isNotEmpty(searchTerm)) {
			return employeeDAO.search(searchTerm, accountId);
		}

		return Collections.emptyList();
	}

	/* All Save Methods */

	@Override
	public Employee save(Employee employee, final EntityAuditInfo entityAuditInfo) {
		employee = EntityHelper.setCreateAuditFields(employee, entityAuditInfo);
		return employeeDAO.save(employee);
	}

	public void save(final Collection<Employee> employees, final EntityAuditInfo entityAuditInfo) {
		EntityHelper.setCreateAuditFields(employees, entityAuditInfo);
		employeeDAO.save(employees);
	}

	/* All Update Methods */

	@Override
	public Employee update(final Employee employee, final EntityAuditInfo entityAuditInfo) {
		Employee employeeToUpdate = find(employee.getId());
		employeeToUpdate = updatePersonalInformation(employeeToUpdate, employee);
		employeeToUpdate = updateEmployeeEmploymentInformation(employeeToUpdate, employee);

		employeeToUpdate = EntityHelper.setUpdateAuditFields(employeeToUpdate, entityAuditInfo);

		return employeeDAO.save(employeeToUpdate);
	}

	private Employee updatePersonalInformation(final Employee employeeToUpdate, final Employee employee) {
		employeeToUpdate.setFirstName(employee.getFirstName());
		employeeToUpdate.setLastName(employee.getLastName());
		employeeToUpdate.setEmail(employee.getEmail());
		employeeToUpdate.setPhone(employee.getPhone());
		return employeeToUpdate;
	}

	private Employee updateEmployeeEmploymentInformation(final Employee employeeToUpdate, final Employee employee) {
		employeeToUpdate.setSlug(generateSlug(employee));
		employeeToUpdate.setPositionName(employee.getPositionName());

		return employeeToUpdate;
	}

	private String generateSlug(final Employee employee) {
		if (Strings.isNotEmpty(employee.getSlug())) {
			return employee.getSlug();
		}

		String hash = Strings.hashUrlSafe(employee.getAccountId() + employee.getEmail());
		return "EID-" + hash.substring(0, 8).toUpperCase();
	}

	/* All Delete Methods */

	@Override
	public void delete(final Employee employee) {
		if (employee == null) {
			throw new NullPointerException("employee cannot be null");
		}

		employeeDAO.delete(employee);
	}

	@Override
	public void deleteById(final Integer id) {
		if (id == null) {
			throw new NullPointerException("id cannot be null");
		}

		Employee employee = find(id);
		delete(employee);
	}

	public void delete(final int id, final int accountId) {
		Employee employee = find(id, accountId);
		delete(employee);
	}

	/* Additional Methods */

	public Employee updatePhoto(final PhotoForm photoForm, final String directory, final int id,
	                            final int accountId) throws Exception {

		String extension = FileUtils.getExtension(photoForm.getPhotoFileName()).toLowerCase();

		if (photoUtil.isValidExtension(extension)) {
			String filename = PICSFileType.employee_photo.filename(id) + "-" + accountId;
			photoUtil.sendPhotoToFilesDirectory(photoForm.getPhoto(), directory, id, extension, filename);
		} else {
			throw new IllegalArgumentException("Invalid file format");
		}

		return find(id, accountId);
	}
}
