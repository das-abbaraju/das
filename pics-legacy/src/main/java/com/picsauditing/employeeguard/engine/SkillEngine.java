package com.picsauditing.employeeguard.engine;

import com.picsauditing.employeeguard.entities.AccountSkill;
import com.picsauditing.employeeguard.entities.Employee;
import com.picsauditing.employeeguard.entities.Project;
import com.picsauditing.employeeguard.entities.Role;
import com.picsauditing.employeeguard.services.AccountService;
import com.picsauditing.employeeguard.services.entity.*;
import com.picsauditing.employeeguard.services.models.AccountModel;
import com.picsauditing.employeeguard.util.PicsCollectionUtil;
import org.apache.commons.collections.CollectionUtils;
import org.apache.commons.collections.MapUtils;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.*;

public class SkillEngine {

	@Autowired
	private AccountService accountService;
	@Autowired
	private EmployeeEntityService employeeEntityService;
	@Autowired
	private GroupEntityService groupEntityService;
	@Autowired
	private ProjectEntityService projectEntityService;
	@Autowired
	private RoleEntityService roleEntityService;
	@Autowired
	private SkillEntityService skillEntityService;

	public Map<Integer, Set<AccountSkill>> getEmployeeSkillsForSites(final Collection<AccountModel> sites,
																	 final Map<Integer, Set<Employee>> siteAssignments,
																	 final Map<Integer, Employee> contractorEmployee) {
//		if (CollectionUtils.isEmpty(sites) || CollectionUtils.isEmpty(employees)) {
//			return Collections.emptyMap();
//		}
//
//		Map<Integer, Set<AccountSkill>> employeeSiteSkills = new HashMap<>();
//		for (Integer siteId : sites) {
//			employeeSiteSkills.put(siteId, getEmployeeSkillsForSite(siteId, employees));
//		}
//
//		return employeeSiteSkills;

		return Collections.emptyMap();
	}

	public Map<Employee, Set<AccountSkill>> getEmployeeSkillsMapForAccount(final Collection<Employee> employees,
																		   final AccountModel accountModel) {
		Map<Employee, Set<AccountSkill>> employeeGroupSkills =
				skillEntityService.getGroupSkillsForEmployees(groupEntityService.getEmployeeGroups(employees));
		Map<Employee, Set<AccountSkill>> employeeRoleSkills = getRoleSkillsForEmployees(employees, accountModel);
		Map<Employee, Set<AccountSkill>> employeeProjectSkills = getProjectSkillsForEmployees(employees, accountModel);

		Map<Employee, Set<AccountSkill>> employeeGroupRoleSkills = PicsCollectionUtil.mergeMapOfSets(employeeGroupSkills, employeeRoleSkills);
		return PicsCollectionUtil.mergeMapOfSets(employeeGroupRoleSkills, employeeProjectSkills);
	}

	private Map<Employee, Set<AccountSkill>> getProjectSkillsForEmployees(final Collection<Employee> employees,
																		  final AccountModel accountModel) {
		Map<Employee, Set<Project>> employeeProjects = getProjectSkills(employees, accountModel);
		Map<Project, Set<AccountSkill>> projectRequiredSkills =
				skillEntityService.getRequiredSkillsForProjects(PicsCollectionUtil.extractAndFlattenValuesFromMap(employeeProjects));

		return getEmployeeSkillsMap(employees, employeeProjects, projectRequiredSkills);
	}

	private Map<Employee, Set<AccountSkill>> getRoleSkillsForEmployees(Collection<Employee> employees, AccountModel accountModel) {
		Map<Employee, Set<Role>> employeeRoles = getEmployeeSiteAssignments(employees, accountModel);
		Map<Role, Set<AccountSkill>> roleRequiredSkills =
				skillEntityService.getSkillsForRoles(PicsCollectionUtil.extractAndFlattenValuesFromMap(employeeRoles));

		return getEmployeeSkillsMap(employees, employeeRoles, roleRequiredSkills);
	}

	private <ENTITY> Map<Employee, Set<AccountSkill>> getEmployeeSkillsMap(final Collection<Employee> employees,
																		   final Map<Employee, Set<ENTITY>> employeeToEntities,
																		   final Map<ENTITY, Set<AccountSkill>> entitySkills) {
		if (MapUtils.isEmpty(employeeToEntities) || MapUtils.isEmpty(entitySkills)) {
			return Collections.emptyMap();
		}

		Map<Employee, Set<AccountSkill>> employeeSkills = new HashMap<>();
		for (Employee employee : employees) {
			if (!employeeSkills.containsKey(employee)) {
				employeeSkills.put(employee, new HashSet<AccountSkill>());
			}

			if (employeeToEntities.containsKey(employee)) {
				for (ENTITY entity : employeeToEntities.get(employee)) {
					if (CollectionUtils.isNotEmpty(entitySkills.get(entity))) {
						employeeSkills.get(employee).addAll(entitySkills.get(entity));
					}
				}
			}
		}

		return employeeSkills;
	}

	private Map<Employee, Set<Project>> getProjectSkills(final Collection<Employee> employees,
														 final AccountModel accountModel) {
		switch (accountModel.getAccountType()) {

			case CONTRACTOR:
				return projectEntityService.getProjectsForEmployees(employees);

			case CORPORATE:
				Collection<Integer> siteIds = accountService.getChildOperatorIds(accountModel.getId());
				return projectEntityService.getProjectsForEmployeesBySiteIds(employees, siteIds);

			case OPERATOR:
				return projectEntityService.getProjectsForEmployeesBySiteId(employees, accountModel.getId());

			default:
				throw new IllegalArgumentException("Invalid account type");
		}
	}

	private Map<Employee, Set<Role>> getEmployeeSiteAssignments(final Collection<Employee> employees,
																final AccountModel accountModel) {
		int id = accountModel.getId();

		switch (accountModel.getAccountType()) {
			case CONTRACTOR:
				return getEmployeesToRolesForContractor(id);

			case CORPORATE:
				Collection<Integer> siteIds = accountService.getChildOperatorIds(id);
				return roleEntityService.getProjectRolesForEmployees(employees, siteIds);

			case OPERATOR:
				return roleEntityService.getProjectRolesForEmployees(employees, id);

			default:
				throw new IllegalArgumentException("Invalid account type");
		}
	}

	private Map<Employee, Set<Role>> getEmployeesToRolesForContractor(int id) {
		List<Integer> siteIds = accountService.getOperatorIdsForContractor(id);
		Map<Role, Set<Employee>> rolesToEmployees = employeeEntityService.getEmployeesBySiteRoles(siteIds);
		Collection<Project> projects = projectEntityService.getProjectsBySiteIds(siteIds);
		Map<Role, Set<Employee>> projectRolesToEmployees = employeeEntityService.getEmployeesByProjectRoles(projects);

		Map<Role, Set<Employee>> allRolesToEmployees = PicsCollectionUtil.mergeMapOfSets(rolesToEmployees, projectRolesToEmployees);
		return PicsCollectionUtil.invertMapOfSet(allRolesToEmployees);
	}
}
