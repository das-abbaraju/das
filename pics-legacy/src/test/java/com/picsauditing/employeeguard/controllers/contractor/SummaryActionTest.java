package com.picsauditing.employeeguard.controllers.contractor;

import com.picsauditing.PicsActionTest;
import com.picsauditing.actions.PicsActionSupport;
import com.picsauditing.employeeguard.engine.SkillEngine;
import com.picsauditing.employeeguard.entities.AccountSkill;
import com.picsauditing.employeeguard.entities.Employee;
import com.picsauditing.employeeguard.entities.builders.AccountSkillBuilder;
import com.picsauditing.employeeguard.entities.builders.EmployeeBuilder;
import com.picsauditing.employeeguard.services.AccountService;
import com.picsauditing.employeeguard.services.StatusCalculatorService;
import com.picsauditing.employeeguard.services.calculator.SkillStatus;
import com.picsauditing.employeeguard.services.entity.EmployeeEntityService;
import com.picsauditing.employeeguard.services.models.AccountModel;
import org.approvaltests.Approvals;
import org.approvaltests.reporters.DiffReporter;
import org.approvaltests.reporters.UseReporter;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.powermock.reflect.Whitebox;

import java.util.*;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.mockito.Matchers.*;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@UseReporter(DiffReporter.class)
public class SummaryActionTest extends PicsActionTest {

	public static final int SITE_ID = 123;
	private SummaryAction summaryAction;

	@Mock
	private AccountService accountService;
	@Mock
	private EmployeeEntityService employeeEntityService;
	@Mock
	private SkillEngine skillEngine;
	@Mock
	private StatusCalculatorService statusCalculatorService;

	@Before
	public void setUp() throws Exception {
		MockitoAnnotations.initMocks(this);

		summaryAction = new SummaryAction();
		super.setUp(summaryAction);

		when(permissions.getAccountId()).thenReturn(SITE_ID);

		Whitebox.setInternalState(summaryAction, "accountService", accountService);
		Whitebox.setInternalState(summaryAction, "employeeEntityService", employeeEntityService);
		Whitebox.setInternalState(summaryAction, "skillEngine", skillEngine);
		Whitebox.setInternalState(summaryAction, "statusCalculatorService", statusCalculatorService);
	}

	@Test
	public void testIndex() throws Exception {
		setupForIndex();

		assertEquals(PicsActionSupport.JSON_STRING, summaryAction.index());
		performVerificationsForIndex();

		Approvals.verify(summaryAction.getJsonString());
	}

	private void setupForIndex() {
		AccountModel accountModel = new AccountModel.Builder().id(SITE_ID).build();
		final Employee employee = new EmployeeBuilder().build();
		final List<Employee> employees = Arrays.asList(employee);
		final AccountSkill accountSkill = new AccountSkillBuilder().build();
		HashMap<Employee, Set<AccountSkill>> employeeSkills = new HashMap<Employee, Set<AccountSkill>>() {{
			put(employee, new HashSet<AccountSkill>());
			get(employee).add(accountSkill);
		}};
		HashMap<Employee, SkillStatus> employeeStatus = new HashMap<Employee, SkillStatus>() {{
			put(employee, SkillStatus.Expiring);
		}};

		when(accountService.getAccountById(SITE_ID)).thenReturn(accountModel);
		when(employeeEntityService.getEmployeesForAccount(SITE_ID)).thenReturn(employees);
		when(employeeEntityService.getRequestedEmployeeCount(SITE_ID)).thenReturn(4);
		when(skillEngine.getEmployeeSkillsMapForAccount(employees, accountModel)).thenReturn(employeeSkills);
		when(statusCalculatorService.getEmployeeStatusRollUpForSkills(employees, employeeSkills))
				.thenReturn(employeeStatus);
	}

	private void performVerificationsForIndex() {
		assertNotNull(summaryAction.getJsonString());

		verify(accountService).getAccountById(SITE_ID);
		verify(employeeEntityService).getEmployeesForAccount(SITE_ID);
		verify(employeeEntityService).getRequestedEmployeeCount(SITE_ID);
		verify(skillEngine).getEmployeeSkillsMapForAccount(anyListOf(Employee.class), any(AccountModel.class));
		verify(statusCalculatorService).getEmployeeStatusRollUpForSkills(anyCollectionOf(Employee.class), anyMap());
	}
}
