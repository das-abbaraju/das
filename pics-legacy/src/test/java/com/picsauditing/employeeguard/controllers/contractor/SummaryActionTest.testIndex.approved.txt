{"completed":0,"pending":0,"expiring":1,"expired":0,"requested":4}